tools/internal_ci/helper_scripts/delete_leftover_loadtests.sh

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+# Copyright 2021 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -ex
+
+echo "BEGIN Listing and deleting leftover tests."
+
+# Find tests that have running pods and are in Errored state, and delete them.
+kubectl get pods --no-headers -o jsonpath='{range .items[*]}{.metadata.ownerReferences[0].name}{" "}{.status.phase}{"\n"}{end}' \
```

Comment:
1. why do we need this at all? If all loadtest pods reliably self-destruct themselves X minutes after they start (which is something we discussed and it quite easy to achieve), how can we  ever have pods that keep running forever?
2. I though each loadtest has a time-to-live set to sth like 24h, after which it gets deleted automatically by the controller. Does this mean that mechanism is not working reliably.

3. from the command, I can't tell if this "cleanup" phase would potentially also delete pods / loadtests that weren't started by the CI at all, but were actually started manually by someone (e.g. me) when performing a manual experiment. I really wouldn't want the CI job to accidentally delete stuff from my experiment under my hands and I don't see anything in this command that would prevent that from happening.

My concern is that by adding this, we are trying to substitute having a reliable benchmarking harness (where under normal circumstances no "leftover" jobs should exist imho because of 1 and 2.) by periodically running a cleanup script that IMHO can do more harm than good (e.g. delete people's failed experiments under their hands before they have a chance to debug what went wrong. Or it can simply mask that some components in our harness aren't working reliably).

Reply:
Changed the PR to list rather than delete the pods in this condition.

Listing the pods is useful in the long term to monitor for issues (for instance, if the "self-destruct" mechanism doesn't work).

Deleting the pods would have been useful as a workaround, to keep the continuous runs from failing while we put in place and test a mechanism to achieve this "self-destruct" (not only for the driver but for the unresponsive workers, which may well turn out to be unresponsive to the self-destruct signal as well).

Since we are no longer deleting the pods, we have two options to keep the lights on in the meantime:

a. Disable the offending test. I am sending out https://github.com/grpc/grpc/pull/26339 to achieve that.
b. Increase the capacity of our node pools to support 20 simultaneous tests instead of 8 (we currently run only 4 tests simultaneously on the two kokoro jobs, and the other 4 were set aside for issues such as this – too few, as it turns out).

The automatic deletion of tests (and pods) after 24h works reliably. The timeout of 15 minutes is also respected by the test, which marks itself as Errored after this time. The problem is in the interaction of the driver and workers, in the particular case where a worker becomes unresponsive. The driver sends messages to the workers telling them to quit. If the workers are unresponsive, they keep running. The driver appears to wait for a response forever, and not quit either, so the driver and worker pods stay in Running state forever (or until they are deleted, at the end of 24 hours).

I will apply option (b) if option (a) is not preferred. We will then work on the implementation of the "self-destruct".

In the meantime, I would like to merge this PR to enable monitoring in case other tests misbehave.

Reply:
I don't see why the "self-destruct" mechanism would be unreliable if done right.
e.g. when starting the pod, before you start any driver or worker process, you can start a background shell script that waits for given timeout and after that emits a `kill -9` (SIGKILL) to hard kill the driver / worker process. I have hard time imagining circumstances where kill -9 wouldn't work (it certainly would work for killing qps_worker or driver regardless of how "badly" they got stuck).
The mechanism needs to be in place for all the pods we run (not just the driver, but also the workers - the workers are in a sense more likely to get stuck since they are component that gets stressed the most when the benchmark runs).